teps/0101-env-in-pod-template.md

```diff
@@ -0,0 +1,642 @@
+Kubernetes ETCD limitation. 
+
+Besides, users quite frequently are willing to overwrite environment variable values
+specified in a `stepTemplate` in the single place when running pipelines.
+That helps to optionally pass settings, without a need to define additional pipeline parameters.
+
+Having an `env` field in the pod template allows to:
+
+- specify global level defaults, what is important to reduce the size of `TaskRun` and `PipelineRun`
+- override defaults from `stepTemplate` at `TaskRun` and `PipelineRun` level
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+1. The main goal of this proposal is to enable support for specification of environment variables on the global level
+(`TaskRun` and `PipelineRun`).
+
+2. Environment variables defined in the Pod template at `TaskRun` and `PipelineRun` level
+   take precedence over ones defined in steps and `stepTemplate`.
+   
+3. Allow cluster admin to define a list of cluster-wide forbidden environment variables so that users won't overwritten
+   important Tekton environment variables such as "HTTP_PROXY".
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+### Use Cases
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+1. In the first case, common environment variables can be defined in a single place on a `PipelineRun` level.
+   Values can be specified as literals or through Kubernetes references.
+   Variables defined on a `PipelineRun` or `TaskRun` level are then available in all steps.
+   That allows to simplify the Tekton run resource configuration and significantly reduce the size of
+   `PipelineRun` and `TaskRun` resource,
+   by excluding the common environment variables like: static global settings, common values coming from metadata, ...
+
+2. Secondly, environment variables defined in steps can be easily overwritten by the ones from `PipelineRun` and `TaskRun`.
+  With that, common settings like API keys, connection details, ... can be optionally overwritten in a single place.
+  
+3. For Cloud Providers, it's very common to inject user credentials using Kubernetes API `valueFrom` to avoid credentials
+   being exposed to the PodSpec. Since each cloud provider has different credential format, able to assign environment 
+   variables at the `PipelineRun` and `TaskRun` can reuse the same task with different Cloud Provider credentials.
+   Kubernetes API `valueFrom` can also refe to values in the pod labels/annotations for specific Kubernetes cluster
+   information such as namespace, application labels, and service annotations.
```

Comment:
Currently we can use mutating webhook in operator to add the proxy envs in the taskrun pod at pod level, this can be easily eliminated after this feature will be provided in pipeline. So a common use case can be to have proxy support in tekoncd/pipeline and proxy env vars can be defined in global tektoncd/pipeline level default